### PR TITLE
dataflow: new minimal gpu samples

### PR DESCRIPTION
**Pull request by davidcavazos**
## Description

- Adds new minimal examples to use GPU with tensorflow and pytorch.
- Simplifies the landsat sample with newer patterns.
- Fixes some resource leaks due to uncaught errors in some fixture's teardown code.

## Checklist
- [ ] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/AUTHORING_GUIDE.md)
- [ ] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/AUTHORING_GUIDE.md#readme-file)
- [ ] **Tests** pass:   `nox -s py-3.6` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] Please **merge** this PR for me once it is approved.
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/.github/CODEOWNERS) with the codeowners for this sample

### CODE REVIEW

**Comment by kurtisvg** [comment]
@davidcavazos Would you mind pinging me for review once you have the tests passing?

File: dataflow/gpu-workers/pytorch-minimal/README.md
Comment (tvalentyn) [inline dataflow/gpu-workers/pytorch-minimal/README.md:30]: We no longer need to  specify the zone, zone adviser will pick the best zone for the user.

File: dataflow/gpu-workers/pytorch-minimal/README.md
Comment (tvalentyn) [inline dataflow/gpu-workers/pytorch-minimal/README.md:39]: Accordingly, GPUs should be available in some zones in this region.

File: dataflow/gpu-workers/tensorflow-landsat/main.py
Comment (tvalentyn) [inline dataflow/gpu-workers/tensorflow-landsat/main.py:281]: I think users may not understand this comment, let's remove it. Re `with` behavior, that may be WAI - for example when I access files in a `with` statement, resources are cleaned up after we exit from the context, so it makes sense to me that the `with` statement waits for pipeline completion.

File: dataflow/gpu-workers/tensorflow-landsat/run.yaml
Comment (tvalentyn) [inline dataflow/gpu-workers/tensorflow-landsat/run.yaml:31]: Same here (no need to hardcode a zone)

File: dataflow/gpu-workers/tensorflow-landsat/main.py
Comment (tvalentyn) [inline dataflow/gpu-workers/tensorflow-landsat/main.py:287]: How abount:

We pass `gpu_check` as an unused side input to make a blocking one-time check
in the pipeline to verify GPUs are available.

File: dataflow/gpu-workers/pytorch-minimal/main.py
Comment (tvalentyn) [inline dataflow/gpu-workers/pytorch-minimal/main.py:44]: Maybe add a comment: See landsat example to make this a one-time check instead of a per-element check.

Comment (tvalentyn) [reply dataflow/gpu-workers/pytorch-minimal/main.py:44]: Or perhaps just make it a parallel check if you think it's not too complex - one of CE's suggested that would be helpful - leaving this up to you..

File: dataflow/gpu-workers/pytorch-minimal/main.py
Comment (tvalentyn) [inline dataflow/gpu-workers/pytorch-minimal/main.py:45]: We could add a comment that users can sub this with a code that actually uses GPU here or perhaps make it a separate function + add a comment.

File: dataflow/gpu-workers/pytorch-minimal/README.md
Comment (davidcavazos) [reply dataflow/gpu-workers/pytorch-minimal/README.md:30]: Awesome, that's great to know :)

File: dataflow/gpu-workers/pytorch-minimal/README.md
Comment (davidcavazos) [reply dataflow/gpu-workers/pytorch-minimal/README.md]: Thanks, removed this note.

File: dataflow/gpu-workers/tensorflow-landsat/main.py
Comment (davidcavazos) [reply dataflow/gpu-workers/tensorflow-landsat/main.py:278]: Alright, I'm removing this comment for now.

File: dataflow/gpu-workers/tensorflow-landsat/run.yaml
Comment (davidcavazos) [reply dataflow/gpu-workers/tensorflow-landsat/run.yaml]: Removed

File: dataflow/gpu-workers/tensorflow-landsat/main.py
Comment (davidcavazos) [reply dataflow/gpu-workers/tensorflow-landsat/main.py]: Removed this and used a more compact pattern.

File: dataflow/gpu-workers/pytorch-minimal/main.py
Comment (davidcavazos) [reply dataflow/gpu-workers/pytorch-minimal/main.py]: I used a more compact and copy-paste friendly pattern.

File: dataflow/gpu-workers/tensorflow-minimal/README.md
Comment (tvalentyn) [inline dataflow/gpu-workers/tensorflow-minimal/README.md:1]: Tensorflow

Comment (davidcavazos) [reply dataflow/gpu-workers/tensorflow-minimal/README.md]: Good catch, fixed!

File: dataflow/gpu-workers/pytorch-minimal/Dockerfile
Comment (tvalentyn) [inline dataflow/gpu-workers/pytorch-minimal/Dockerfile:26]: We could make a comment somewhere that since users already install requirements.txt in the dockerfile, they should not pass --requirements_file when they launch the pipeline.

File: dataflow/gpu-workers/conftest.py
Comment (tvalentyn) [inline dataflow/gpu-workers/conftest.py:62]: Do you use some random suffixes here to make sure two tests don't step on each other?

File: dataflow/gpu-workers/conftest.py
Comment (tvalentyn) [inline dataflow/gpu-workers/conftest.py:67]: Does the cleanup code  after `yield`  get executed if we only create one bucket?

File: dataflow/gpu-workers/pytorch-minimal/e2e_test.py
Comment (tvalentyn) [inline dataflow/gpu-workers/pytorch-minimal/e2e_test.py:42]: Does this mean you only run one dataflow job per session? It doesn't seem to make sense as the test suite should run as many jobs as requested, on the other hand it's fine to create only one bucket for test session if you'd like to reuse the same bucket.

File: dataflow/gpu-workers/pytorch-minimal/main.py
Comment (tvalentyn) [inline dataflow/gpu-workers/pytorch-minimal/main.py:57]: no strong opinion  but this might be additional complication to the example. You could as well just create a pcollection with this value.

File: dataflow/gpu-workers/tensorflow-minimal/run.yaml
Comment (tvalentyn) [inline dataflow/gpu-workers/tensorflow-minimal/run.yaml:47]: since this is using tensorflow, let's add:
--experiment=no_use_multiple_sdk_containers

**Review by tvalentyn — APPROVED** [review]

File: dataflow/flex-templates/streaming_beam/requirements.txt
Comment (davidcavazos) [reply dataflow/flex-templates/streaming_beam/requirements.txt]: From a user perspective, Beam is a requirement, so it should be here. If Beam is already installed, reinstalling it should be a no-op.

File: dataflow/gpu-workers/pytorch-minimal/Dockerfile
Comment (davidcavazos) [reply dataflow/gpu-workers/pytorch-minimal/Dockerfile]: Thanks, updated in all. Thankfully we also have a bot to automatically update Docker image versions.

File: dataflow/gpu-workers/pytorch-minimal/Dockerfile
Comment (davidcavazos) [reply dataflow/gpu-workers/pytorch-minimal/Dockerfile:27]: It's always a good idea to upgrade pip. And yes, the pytorch image came with an older version of pip.

If it breaks with a newer version of pip, that sounds like we should file an issue to Beam to fix the dependency version constraints.

File: dataflow/gpu-workers/pytorch-minimal/e2e_test.py
Comment (davidcavazos) [reply dataflow/gpu-workers/pytorch-minimal/e2e_test.py:42]: It's one job for this specific test suite. It doesn't really make any difference since there is only one test that uses this.

In case we wanted to check different things about the job, it would make sense to only launch one job and just poll different properties from it, although in this case we're only waiting until it finishes successfully.

File: dataflow/gpu-workers/pytorch-minimal/main.py
Comment (davidcavazos) [reply dataflow/gpu-workers/pytorch-minimal/main.py:57]: Yeah, my train of thought was that most pipelines need one or more input options, so it would be easier to modify by having that as a pattern. I also don't have a strong opinion, but I think it's common enough that it would still be useful.

File: dataflow/gpu-workers/tensorflow-minimal/run.yaml
Comment (davidcavazos) [reply dataflow/gpu-workers/tensorflow-minimal/run.yaml:47]: Thanks, added to the landsat example as well.

File: dataflow/flex-templates/streaming_beam/requirements.txt
Comment (davidcavazos) [reply dataflow/flex-templates/streaming_beam/requirements.txt]: I'll split these changes to another PR to unblock this one.

File: dataflow/gpu-workers/conftest.py
Comment (davidcavazos) [reply dataflow/gpu-workers/conftest.py:62]: Yes, we use a unique identifier for each test so each test has its own unique resources.

File: dataflow/gpu-workers/conftest.py
Comment (davidcavazos) [reply dataflow/gpu-workers/conftest.py:67]: Yes, fixtures do cleanup after the yield.

File: dataflow/gpu-workers/pytorch-minimal/e2e_test.py
Comment (davidcavazos) [reply dataflow/gpu-workers/pytorch-minimal/e2e_test.py:42]: Also, if we sent multiple jobs, they would have the same job name so subsequent jobs would fail. For these tests, having exactly one job is what we want. It could be different in other test suites, but each one defines how they run their jobs.

File: dataflow/gpu-workers/tensorflow-minimal/run.yaml
Comment (tvalentyn) [reply dataflow/gpu-workers/tensorflow-minimal/run.yaml:47]: landsat uses a 1 core vm, so it's not necessary. you could remove the custom VM machine type, but as far as I remember, you needed a highmem machine type so you had to set some non-standard machine type anyway.

Comment (tvalentyn) [reply dataflow/gpu-workers/tensorflow-minimal/run.yaml:47]: it wouldn't hurt, but users might get confused thinking that both measures are required.

Comment (tvalentyn) [reply dataflow/gpu-workers/tensorflow-minimal/run.yaml:47]: If the default machine type works for you, you could just use the flag and omit the machine type.

**Comment by davidcavazos** [comment]
Hi @kurtisvg, all tests are passing, this should be ready for review.

File: dataflow/gpu-workers/tensorflow-landsat/README.md
Comment (tvalentyn) [inline dataflow/gpu-workers/tensorflow-landsat/README.md:1]: Can we decouple changes to Landsat example into a separate change, or make a new folder like 'landsat_v2'?

We currently refer to Landsat example in https://cloud.google.com/dataflow/docs/samples/satellite-images-gpus , and I think these instructions will no longer work once this is merged, since Python version is no longer a parameter in this dockerfile.

Alternatively, we should merge this simultaneously with an update to https://cloud.google.com/dataflow/docs/samples/satellite-images-gpus.

Comment (tvalentyn) [reply dataflow/gpu-workers/tensorflow-landsat/README.md:1]: The second option may be easier, but we would need to make and test the website changes first.

File: dataflow/gpu-workers/requirements.txt
Comment (tvalentyn) [inline dataflow/gpu-workers/requirements.txt]: thanks for restoring the example. nit: 2.30.0 here as well

Comment (davidcavazos) [reply dataflow/gpu-workers/requirements.txt]: Good catch! Thanks

**Comment by davidcavazos** [comment]
Restored the old landsat sample to avoid breaking the current docs, and moved the new sample into a new directory because otherwise there were conflicts in the directory pytest ran tests from.

File: dataflow/gpu-examples/pytorch-minimal/README.md
Comment (leahecole) [inline dataflow/gpu-examples/pytorch-minimal/README.md:30]: If it's not necessary to use beta here, let's use regular `gcloud builds submit`

File: dataflow/gpu-examples/tensorflow-landsat/README.md
Comment (leahecole) [inline dataflow/gpu-examples/tensorflow-landsat/README.md:35]: same comment about using `beta`

File: dataflow/gpu-examples/tensorflow-landsat/main.py
Comment (leahecole) [inline dataflow/gpu-examples/tensorflow-landsat/main.py:69]: Thank you for commenting what the locations are!! So cool. 🌍 🌎 🌏

File: dataflow/gpu-examples/tensorflow-minimal/README.md
Comment (leahecole) [inline dataflow/gpu-examples/tensorflow-minimal/README.md:30]: same comment about beta

**Review by leahecole — CHANGES_REQUESTED** [review]
Mostly little nits - I also wanted to check that it was intentional to commit the .gcloudignore and .dockerignore files

File: dataflow/gpu-examples/pytorch-minimal/README.md
Comment (davidcavazos) [reply dataflow/gpu-examples/pytorch-minimal/README.md:30]: I was using beta because only beta supports streaming the logs, so without beta it looked like a black box process that hanged indefinitely. It definitely works without the beta. There is a console message suggesting to use beta if you want logs, so I guess we can leave that to the user.

**Comment by davidcavazos** [comment]
The .gcloudgnore and .dockerignore were part of the sample before, mostly to avoid uploading the Python virtualenv and other files to Cloud Build which were several GiB. However, now we're doing everything in Cloud Build, so we don't even need Python installed locally. I removed them because now they're irrelevant.

We'll remove the older landsat example under gpu-workers once we update the docs.

**Comment by davidcavazos** [comment]
@leahecole @tvalentyn, tests are passing, this should be ready to merge if there are no more comments.

File: dataflow/gpu-examples/conftest.py
Comment (leahecole) [inline dataflow/gpu-examples/conftest.py:185]: Is there ever a circumstance where this process wouldn't be stopped?

File: dataflow/gpu-examples/conftest.py
Comment (leahecole) [inline dataflow/gpu-examples/conftest.py:296]: If we update past v1b3 will this build command fail? (just making sure it's future proofed so that when dependencies get updated we don't break samples)

File: dataflow/gpu-examples/conftest.py
Comment (leahecole) [inline dataflow/gpu-examples/conftest.py:334]: Non blocking nit - what do you think about these also being constants like the other `RETRY_MAX_TIME`? Might help make them more visible for long term maintenance.

File: dataflow/gpu-examples/conftest.py
Comment (leahecole) [inline dataflow/gpu-examples/conftest.py:481]: Okay @davidcavazos in general I'm a fan of your extremely helpful comments but this one is my favorite and I think the most helpful in this file so far 😄

File: dataflow/gpu-examples/pytorch-minimal/e2e_test.py
Comment (leahecole) [inline dataflow/gpu-examples/pytorch-minimal/e2e_test.py:19]: clever!

File: dataflow/gpu-examples/pytorch-minimal/noxfile_config.py
Comment (leahecole) [inline dataflow/gpu-examples/pytorch-minimal/noxfile_config.py:25]: 🎉

File: dataflow/gpu-examples/pytorch-minimal/run.yaml
Comment (leahecole) [inline dataflow/gpu-examples/pytorch-minimal/run.yaml:31]: Does it matter that these substitutions aren't defined? (and/or do they need to be defined in our test project cloud build?)

**Review by leahecole — CHANGES_REQUESTED** [review]
David this is so so so close and also it's a really fantastic PR. When you're back, let's sit down for like 15 min and just knock out my questions. (I sent you an email about this too in case your inbox is gross from being OOO)

File: dataflow/gpu-examples/conftest.py
Comment (davidcavazos) [reply dataflow/gpu-examples/conftest.py:194]: I made some experiments:

* Joining with timeout 0 actually doesn't do anything since it's an infinite loop so it never finishes.
* Terminate always stops the subprocess, but my experiments are linear, and pytest might be doing some nasty concurrency
* If I don't terminate the process, the whole thing hangs because the main process waits for the subprocess to finish, which might explain the intermittent hangs I was getting (if for some reason the fixture cleanup doesn't execute)
* Apparently, we can set up [`daemon`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process.daemon) to True and the main process no longer waits for the subprocess to finish, so it terminates it automatically even if I don't explicitly terminate it :)

I deleted the `join`, kept the `terminate`, but also made the subprocess into a daemon, and added a large explanation on it.

File: dataflow/gpu-examples/conftest.py
Comment (davidcavazos) [reply dataflow/gpu-examples/conftest.py:304]: I don't think it would break, but I don't think it receives a ton of maintenance either. I personally prefer the gcloud commands for these things, but I know the general rule is to avoid them if there's a client library.

File: dataflow/gpu-examples/conftest.py
Comment (davidcavazos) [reply dataflow/gpu-examples/conftest.py]: Sure, sounds good.

On another note, the `RETRY_MAX_TIME` constant is actually not used anymore so I deleted it as well :)

File: dataflow/gpu-examples/pytorch-minimal/run.yaml
Comment (davidcavazos) [reply dataflow/gpu-examples/pytorch-minimal/run.yaml:31]: If they're undefined, Dataflow will find suitable defaults. The README doesn't include to use these, but for the tests we actually define these so we can keep track of the resources.

**Comment by davidcavazos** [comment]
@leahecole the new Beam version depends on `google-cloud-profiler` which requires a C++ compiler and Python build tools in the container image. Also the newer PyTorch image is quite larger so I had to adjust the timeout for building the image.

All tests are passing again.

**Review by leahecole — APPROVED** [review]

**Comment by leahecole** [comment]
Used admin merge just because second approver didn't have correct permissions, but they are a Dataflow expert.

**Pull request merged**
